### DISCUSSION TITLE
Where are the docs & where are the config files?

### DISCUSSION THREAD

**Discussion by txtyash** [woelper/oculante#284] [Q&A] [answered]
Where are the config files? Are they supposed to be under ~/.config? Is it even configurable? No docs?

**Comment by woelper** [comment, marked as answer]
Hi!
> Where are the config files?

Linux	$XDG_DATA_HOME or $HOME/.local/share	/home/alice/.local/share
macOS	$HOME/Library/Application Support	/Users/Alice/Library/Application Support
Windows	{FOLDERID_LocalAppData}	C:\Users\Alice\AppData\Local

plus the filename `.oculante`. 

from: 
https://docs.rs/dirs/5.0.1/dirs/fn.data_local_dir.html

>  Is it even configurable?

Yes, through the UI, in settings. The format is JSON, so you can also edit the file directly.

> No docs?

You can have a look here:
https://github.com/woelper/oculante/blob/9e6eb819a3a44e07cda38e11d22fde3ea67f0795/src/settings.rs#L19-L57

If you prefer to read the docs, you can clone the repo and run `cargo doc --no-deps --open`. You will then be able to see the `PersistentSettings` struct along with the documentation.

**Reply by Gigas002** [reply]
Just to add my 5 cents here, I personally think it'd be better to store configs using `config_dir` or `config_local_dir`. After all, `.config` is a right place for configs. MacOS and Windows users won't face any difference, of the above crate's values are used.

**Reply by woelper** [reply]
Great point, and I think you are right. I'll probably need tome migration strategy, but this seems good. Feel free to open an issue if you like!

**Comment by hrxn** [comment]
@woelper 

> Windows {FOLDERID_LocalAppData} C:\Users\Alice\AppData\Local
> plus the filename `.oculante`

Has this been changed recently?
It seems to be `%LOCALAPPDATA%\oculante\config.json` now, right?

> The format is JSON, so you can also edit the file directly.

I just tried to do that, but somehow it behaves a bit unexpectedly..

https://github.com/woelper/oculante/blob/e468e6bec7438077337041ebc2b50313c5954759/src/settings.rs#L46-L47

Don't really know rust, but u32 is surely 32-bit unsigned integer. I assume those values are in pixels?
Maybe because of serialization from JSON in rust?

**Reply by woelper** [reply]
Yes, this has indeed been changed. It now uses this location:
https://docs.rs/dirs/latest/dirs/fn.config_local_dir.html

This makes it a bit more in line (at least on unix/linux) with how other tools save their configuration.

as for `window_geometry`, the first tuple (position) is ignored as position is not read. The width and height should be though. Do you get different values there if you resize the window?

serde (the serialization framework) makes sure the types can be safely read, even if the format does not support all flavours of data, such as `u32`. I think you can just type something like `[[0,0] , [800,600]]` there. Be aware that this will be overwritten as soon as you change the window size.

**Reply by hrxn** [reply]
> This makes it a bit more in line (at least on unix/linux) with how other tools save their configuration.
>

Not just the Nixes, that's also more in line with how it works on Windows. There are some exceptions, programs who dump their config file etc. directly into `%LOCALAPPDATA%` but that is definitely not the norm.
 
> as for `window_geometry`, the first tuple (position) is ignored as position is not read. The width and height should be though. Do you get different values there if you resize the window?
>

Ah, that would explain it. Width and Height seem to work indeed, but I couldn't figure out how position works so far. Not really surprising if those values are not even read. So, what's happening in the end? Oculante app window randomly picking a position?
 
> serde (the serialization framework) makes sure the types can be safely read, even if the format does not support all flavours of data, such as `u32`. I think you can just type something like `[[0,0] , [800,600]]` there. Be aware that this will be overwritten as soon as you change the window size.

I noticed. I was doing the usual song-and-dance of having a config.json open in my editor, saving after any changes, copy-and-replace-ing the file to the correct place in the terminal and then launching oculante again. At least it works.
I did not test yet if I can set the file to read-only and see if oculante would respect that.

**Reply by woelper** [reply]
> Ah, that would explain it. Width and Height seem to work indeed, but I couldn't figure out how position works so far. Not really surprising if those values are not even read. So, what's happening in the end? Oculante app window randomly picking a position?

Oculante is doing it like most other applications and lets the OS/Window manager assign a position. I have thought long about saving the position, but issues like opening it on an external monitor, closing Oculante and disconnecting the monitor could lead to a position that is off screen. For that reason I decided to keep it simple.

> I did not test yet if I can set the file to read-only and see if oculante would respect that.

Good question, I'd have to look at the code. I think opening would work, and there should never be a panic if it can't write. I _think_ what would happen is that it tries to write, but won't succeed. Maybe I could add a "only save configuration manually" option in the future?

**Reply by hrxn** [reply]
> Oculante is doing it like most other applications and lets the OS/Window manager assign a position. 

I see. But this behavior (n.b. have not tried `v0.8.15` yet, but I don't think the commit log indicates any changes in that regard) where you can define a size but not a position leads to a specific problem, at least on Windows.
Simple steps to reproduce (delete config first):
1. First launch, app window appears upper-left-ish at somewhat small size.
2. Reposition a bit to the left, increase the width of the app window to the right screen border and close the window.
3. Launch oculante again, the window bar controls on the right are now off screen.

> I have thought long about saving the position, but issues like opening it on an external monitor, closing Oculante and disconnecting the monitor could lead to a position that is off screen. For that reason I decided to keep it simple.
> 

Simple is very good. But speaking a bit off-the-cuff here, if you have Windows with multiple monitors, there is still the "virtual" desktop and the `0, 0` position should always be anchored in the very top-left corner of the total displayed surface. This should always work, and prevent any off screen positioning, I think.
Of course, on macOS `0, 0` might be bottom-left instead of top-left, if I remember something correctly here, but I am not sure at all.  

 
> Good question, I'd have to look at the code. I think opening would work, and there should never be a panic if it can't write. I _think_ what would happen is that it tries to write, but won't succeed. Maybe I could add a "only save configuration manually" option in the future?

Can't speak for everyone, but personally I would really appreciate such an option.

**Reply by hrxn** [reply]
FWIW, after doing the STR above, and then opening `config.json` again I saw this:
```json
  "window_geometry":[
    [
      4294967276,
      57
    ],
    [
      1869,
      645
    ]
  ],
```
I don't know what the first tuple is meant to represent, but that is what has been written into the config file by oculante.

**Reply by woelper** [reply]
Definitely good that I don't try to apply these values :) - That looks like a veeeery wide monitor or a bug, which is more likely...

**Reply by hrxn** [reply]
😄 
Just at-me here to hit me up whenever you want to test me something with Oculante on Windows™
